rosbag2_performance/rosbag2_performance_benchmarking/src/benchmark_publishers.cpp

```diff
@@ -55,44 +54,70 @@ class BenchmarkPublishers : public rclcpp::Node
+  {
+    const auto configurations = config_utils::publisher_groups_from_node_parameters(*this);
+
+    if (configurations.empty()) {
+      RCLCPP_ERROR(get_logger(), "No publishers/producers found in node parameters");
+      return;
+    }
+
+    const std::string node_name(get_fully_qualified_name());
+
+    for (auto & config : configurations) {
+      for (unsigned int i = 0; i < config.count; ++i) {
+        const std::string topic = node_name + "/" + config.topic_root + "_" + std::to_string(i + 1);
+        auto producer = create_benchmark_producer(topic, config);
+        producers_.push_back(producer);
+      }
+    }
+  }
+
+  std::shared_ptr<BenchmarkProducer> create_benchmark_producer(
+    std::string topic,
+    const PublisherGroupConfig & config)
+  {
+    const auto & pConfig = config.producer_config;
+    auto producer = std::make_shared<BenchmarkProducer>();
+
+    producer->producer = msg_utils::create(pConfig.message_type, *this, topic, config);
+    producer->max_messages = pConfig.max_count;
+    producer->period = std::chrono::milliseconds(pConfig.frequency ? 1000 / pConfig.frequency : 1);
+
+    return producer;
+  }
+
+  void producer_job(std::shared_ptr<BenchmarkProducer> producer)
   {
-    const std::string topic_prefix(this->get_fully_qualified_name());
-    auto wait_for_subs = config_utils::wait_for_subscriptions_from_node_parameters(*this);
-
-    for (auto & c : configurations_) {
-      for (unsigned int i = 0; i < c.count; ++i) {
-        auto topic = topic_prefix + "/" + c.topic_root + "_" + std::to_string(i + 1);
-        auto pub = this->create_publisher<std_msgs::msg::ByteMultiArray>(topic, c.qos);
-        publishers_.push_back(pub);
-        producers_.push_back(
-          std::make_unique<ByteProducer>(
-            c.producer_config,
-            [this, topic, wait_for_subs] {
-              if (!wait_for_subs) {return;}
-              if (!rclcpp::ok()) {return;}
-              const double max_subscription_wait_time = 5.0;
-              auto start_time = std::chrono::high_resolution_clock::now();
-              while (!this->count_subscribers(topic)) {
-                std::this_thread::sleep_for(std::chrono::milliseconds(50));
-                auto current_time = std::chrono::high_resolution_clock::now();
-                std::chrono::duration<double> elapsed = current_time - start_time;
-                if (elapsed.count() >= max_subscription_wait_time) {
-                  throw std::runtime_error("Waited too long for " + topic);
-                }
-              }
-            },
-            [pub](std::shared_ptr<std_msgs::msg::ByteMultiArray> msg) {
-              pub->publish(*msg);
-            },
-            [] { /* empty lambda */}));
+    for (auto i = 0u; i < producer->max_messages; ++i) {
+      if (!rclcpp::ok()) {
+        break;
       }
+      std::this_thread::sleep_for(producer->period);
+      producer->producer->produce();
+      ++producer->produced_messages;
```

Comment:
It would be nice to adhere to data encapsulation concept and wrap it in 
```
BenchmarkProducer::produce()
{
    this->msg_producer->produce();
    ++this->produced_messages;
}
```

Reply:
done